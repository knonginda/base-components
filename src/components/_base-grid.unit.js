import BaseGrid from './_base-grid'

describe('@components/_base-grid', () => {
  it('renders its content', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders its content when order by DESC sort by ID', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: 'id',
          order: 'desc',
        },
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders its content by click header', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: 'lastName',
          order: 'asc',
        },
      },
    })
    wrapper
      .findAll('.table > thead > tr > th')
      .at(0)
      .trigger('click')
    expect(wrapper).toMatchSnapshot()
  })

  it('renders its content when filter by DD', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: 'DD',
          sort: '',
          order: '',
        },
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('should select item by click checkbox', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
        selector: true,
      },
    })
    const tr = wrapper.findAll('.table > tbody > tr').at(0)
    const td = tr.findAll('td').at(0)
    const checkbox = td.find('input[type="checkbox"]')
    checkbox.trigger('click')
    expect(gridData[0].isSelected).toBe(true)
  })

  it('should unselect item by click checkbox', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
        selector: true,
      },
    })
    wrapper.setData({
      selectedItems: [
        {
          lastName: 'AA',
          firstName: 'AAA',
          id: 1,
          isSelected: true,
        },
      ],
    })
    const tr = wrapper.findAll('.table > tbody > tr').at(0)
    const td = tr.findAll('td').at(0)
    const checkbox = td.find('input[type="checkbox"]')
    checkbox.trigger('click')
    expect(gridData[0].isSelected).toBe(false)
  })

  it('should select all item by click checkbox', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
        selector: true,
      },
    })
    const th = wrapper.findAll('.table > thead > tr > th').at(0)
    const checkbox = th.find('input[type="checkbox"]')
    checkbox.setChecked(true)
    expect(wrapper.vm.selectedItems.length).toEqual(26)
  })

  it('should unselect all item by click checkbox', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
        selector: true,
      },
    })
    wrapper.setData({
      selectedItems: gridData,
      isSelectAll: true,
    })
    const th = wrapper.findAll('.table > thead > tr > th').at(0)
    const checkbox = th.find('input[type="checkbox"]')
    checkbox.setChecked(false)
    expect(wrapper.vm.selectedItems.length).toEqual(0)
  })

  it('should jump to page 5 when you click on the page 5', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
      },
    })
    const pageItem = wrapper.findAll('.pagination .pageItem').at(4)
    const pageLink = pageItem.find('.pageLink')
    pageLink.trigger('click')
    expect(wrapper.vm.query.offset).toBe(8)
  })

  it('should turn to next page when click next button', () => {
    const wrapper = mount(BaseGrid, {
      propsData: {
        data: gridData,
        columns: gridColumns,
        query: {
          limit: 2,
          offset: 0,
          filterKey: '',
          sort: '',
          order: '',
        },
      },
    })
    const pageItem = wrapper.findAll('.pagination .pageItem').at(9)
    pageItem.trigger('click')
    expect(wrapper.vm.query.offset).toBe(2)
  })
})

let gridColumns = [
  {
    key: 'lastName',
    label: 'Last Name',
  },
  {
    key: 'firstName',
    label: 'First Name',
  },
  {
    key: 'id',
    label: 'ID',
  },
]
let gridData = [
  {
    lastName: 'AA',
    firstName: 'AAA',
    id: 1,
  },
  {
    lastName: 'BB',
    firstName: 'BBB',
    id: 2,
  },
  {
    lastName: 'CC',
    firstName: 'CCC',
    id: 3,
  },
  {
    lastName: 'DD',
    firstName: 'DDD',
    id: 4,
  },
  {
    lastName: 'EE',
    firstName: 'EEE',
    id: 5,
  },
  {
    lastName: 'FF',
    firstName: 'FFF',
    id: 6,
  },
  {
    lastName: 'GG',
    firstName: 'GGG',
    id: 7,
  },
  {
    lastName: 'HH',
    firstName: 'HHH',
    id: 8,
  },
  {
    lastName: 'II',
    firstName: 'III',
    id: 9,
  },
  {
    lastName: 'JJ',
    firstName: 'JJJ',
    id: 10,
  },
  {
    lastName: 'KK',
    firstName: 'KKK',
    id: 11,
  },
  {
    lastName: 'LL',
    firstName: 'LLL',
    id: 12,
  },
  {
    lastName: 'MM',
    firstName: 'MMM',
    id: 13,
  },
  {
    lastName: 'NN',
    firstName: 'NNN',
    id: 14,
  },
  {
    lastName: 'OO',
    firstName: 'OOO',
    id: 15,
  },
  {
    lastName: 'PP',
    firstName: 'PPP',
    id: 16,
  },
  {
    lastName: 'QQ',
    firstName: 'QQQ',
    id: 17,
  },
  {
    lastName: 'RR',
    firstName: 'RRR',
    id: 18,
  },
  {
    lastName: 'SS',
    firstName: 'SSS',
    id: 19,
  },
  {
    lastName: 'TT',
    firstName: 'TTT',
    id: 20,
  },
  {
    lastName: 'UU',
    firstName: 'UUU',
    id: 21,
  },
  {
    lastName: 'VV',
    firstName: 'VVV',
    id: 22,
  },
  {
    lastName: 'WW',
    firstName: 'WWW',
    id: 23,
  },
  {
    lastName: 'XX',
    firstName: 'XXX',
    id: 24,
  },
  {
    lastName: 'YY',
    firstName: 'YYY',
    id: 25,
  },
  {
    lastName: 'ZZ',
    firstName: 'ZZZ',
    id: 26,
  },
]
