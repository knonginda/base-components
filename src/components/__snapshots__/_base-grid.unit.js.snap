// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/_base-grid renders its content 1`] = `
<div class="tableContainer">
  <table class="table">
    <thead>
      <tr>
        <!---->
        <th class="">
          Last Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          First Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          ID
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <!---->
        <td>CC</td>
        <td>CCC</td>
        <td>3</td>
      </tr>
      <tr>
        <!---->
        <td>DD</td>
        <td>DDD</td>
        <td>4</td>
      </tr>
    </tbody>
  </table>
  <div class="bottomSpace">
    <div class="total">
      Total:
      <strong>26</strong></div>
    <ul name="Pagination" class="pagination">
      <li class="pageItem"><a href="#" class="pageLink">
          <baseicon-stub name="left-arrow" width="12" height="12" color="#333333" click="() => {}"></baseicon-stub>
        </a></li>
      <li class="pageItem"><a href="#" class="pageLink">1</a></li>
      <li class="pageItem currentPage"><a href="#" class="pageLink">2</a></li>
      <li class="pageItem"><a href="#" class="pageLink">3</a></li>
      <li class="pageItem"><a href="#" class="pageLink">4</a></li>
      <li class="pageItem"><a href="#" class="pageLink">5</a></li>
      <li class="pageItem"><a href="#" class="pageLink">6</a></li>
      <li class="pageItem"><a href="#" class="pageLink">7</a></li>
      <li class="pageItem"><a class="pageLink">...</a></li>
      <li class="pageItem"><a href="#" class="pageLink">13</a></li>
      <li class="pageItem"><a href="#" class="pageLink">
          <baseicon-stub name="right-arrow" width="12" height="12" color="#333333" click="() => {}"></baseicon-stub>
        </a></li>
    </ul>
  </div>
</div>
`;

exports[`@components/_base-grid renders its content by click header 1`] = `
<div class="tableContainer">
  <table class="table">
    <thead>
      <tr>
        <!---->
        <th class="active">
          Last Name
          <baseicon-stub name="down-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          First Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          ID
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <!---->
        <td>ZZ</td>
        <td>ZZZ</td>
        <td>26</td>
      </tr>
      <tr>
        <!---->
        <td>YY</td>
        <td>YYY</td>
        <td>25</td>
      </tr>
    </tbody>
  </table>
  <div class="bottomSpace">
    <div class="total">
      Total:
      <strong>26</strong></div>
    <ul name="Pagination" class="pagination">
      <!---->
      <li class="pageItem currentPage"><a href="#" class="pageLink">1</a></li>
      <li class="pageItem"><a href="#" class="pageLink">2</a></li>
      <li class="pageItem"><a href="#" class="pageLink">3</a></li>
      <li class="pageItem"><a href="#" class="pageLink">4</a></li>
      <li class="pageItem"><a href="#" class="pageLink">5</a></li>
      <li class="pageItem"><a href="#" class="pageLink">6</a></li>
      <li class="pageItem"><a href="#" class="pageLink">7</a></li>
      <li class="pageItem"><a class="pageLink">...</a></li>
      <li class="pageItem"><a href="#" class="pageLink">13</a></li>
      <li class="pageItem"><a href="#" class="pageLink">
          <baseicon-stub name="right-arrow" width="12" height="12" color="#333333" click="() => {}"></baseicon-stub>
        </a></li>
    </ul>
  </div>
</div>
`;

exports[`@components/_base-grid renders its content when filter by DD 1`] = `
<div class="tableContainer">
  <table class="table">
    <thead>
      <tr>
        <!---->
        <th class="">
          Last Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          First Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          ID
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <!---->
        <td>DD</td>
        <td>DDD</td>
        <td>4</td>
      </tr>
    </tbody>
  </table>
  <div class="bottomSpace">
    <div class="total">
      Total:
      <strong>1</strong></div>
    <ul name="Pagination" class="pagination">
      <!---->
      <li class="pageItem currentPage"><a href="#" class="pageLink">1</a></li>
      <!---->
    </ul>
  </div>
</div>
`;

exports[`@components/_base-grid renders its content when order by DESC sort by ID 1`] = `
<div class="tableContainer">
  <table class="table">
    <thead>
      <tr>
        <!---->
        <th class="">
          Last Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="">
          First Name
          <baseicon-stub name="up-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
        <th class="active">
          ID
          <baseicon-stub name="down-arrow" width="12" height="12" color="#333333" classname="sortIcon" click="() => {}"></baseicon-stub>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <!---->
        <td>ZZ</td>
        <td>ZZZ</td>
        <td>26</td>
      </tr>
      <tr>
        <!---->
        <td>YY</td>
        <td>YYY</td>
        <td>25</td>
      </tr>
    </tbody>
  </table>
  <div class="bottomSpace">
    <div class="total">
      Total:
      <strong>26</strong></div>
    <ul name="Pagination" class="pagination">
      <!---->
      <li class="pageItem currentPage"><a href="#" class="pageLink">1</a></li>
      <li class="pageItem"><a href="#" class="pageLink">2</a></li>
      <li class="pageItem"><a href="#" class="pageLink">3</a></li>
      <li class="pageItem"><a href="#" class="pageLink">4</a></li>
      <li class="pageItem"><a href="#" class="pageLink">5</a></li>
      <li class="pageItem"><a href="#" class="pageLink">6</a></li>
      <li class="pageItem"><a href="#" class="pageLink">7</a></li>
      <li class="pageItem"><a class="pageLink">...</a></li>
      <li class="pageItem"><a href="#" class="pageLink">13</a></li>
      <li class="pageItem"><a href="#" class="pageLink">
          <baseicon-stub name="right-arrow" width="12" height="12" color="#333333" click="() => {}"></baseicon-stub>
        </a></li>
    </ul>
  </div>
</div>
`;
